include/exadg/operators/solution_transfer.h

```diff
@@ -0,0 +1,147 @@
+/*  ______________________________________________________________________
+ *
+ *  ExaDG - High-Order Discontinuous Galerkin for the Exa-Scale
+ *
+ *  Copyright (C) 2023 by the ExaDG authors
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *  ______________________________________________________________________
+ */
+
+#ifndef INCLUDE_EXADG_OPERATORS_SOLUTION_TRANSFER_H
+#define INCLUDE_EXADG_OPERATORS_SOLUTION_TRANSFER_H
+
+// deal.II
+#include <deal.II/distributed/solution_transfer.h>
+#include <deal.II/distributed/tria.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/numerics/solution_transfer.h>
+
+namespace ExaDG
+{
+template<int dim, typename VectorType>
+class SolutionTransfer
+{
+public:
+  /*
+   * Constructor.
+   */
+  SolutionTransfer(dealii::DoFHandler<dim> const & dof_handler_in)
+  {
+    dof_handler = &dof_handler_in;
+  }
+
+  void
+  prepare_coarsening_and_refinement(std::vector<VectorType *> & vectors)
+  {
+    // Container vectors_old_grid hold vectors for interpolation *REQUIRED* for
+    // interpolate_after_coarsening_and_refinement. Thus makes an actual copy.
+    // In the case of parallel::distributed::Triangulation, pointers are sufficient.
+    vectors_old_grid.resize(vectors.size());
+    std::vector<VectorType const *> vectors_old_grid_ptr(vectors.size());
+    for(unsigned int i = 0; i < vectors.size(); ++i)
+    {
+      vectors[i]->update_ghost_values();
+
+      if(is_parallel_distributed_triangulation())
+      {
+        vectors_old_grid_ptr[i] = vectors[i];
+      }
+      else
```

Comment:
Is this branch also relevant for serial triangulations? I am not that familiar with serial, but `locally_owned_elements()`/`get_mpi_communicator()` looks somewhat strange in this context?

Reply:
Yes, it also applies to serial triangulations. The crucial difference is type of solution transfer object used in lines 78 or 85 below, where one needs vectors to stay alive during the transfer (the non-`is_parallel_distributed_triangulation` case) or if the data is stored inside the transfer class somehow.